File: 3000/LC 2044.cpp
class Solution {
public:
    int countMaxOrSubsets(vector<int>& nums) {
        int n = nums.size();
        int cnt = 0, maxVal = 0;
        function<void(int,int)> dfs = [&] (int idx, int orVal) {
            if (idx == n) {
                if (orVal > maxVal) {
                    maxVal = orVal;
                    cnt = 1;
                } else if (orVal == maxVal) {
                    cnt++;
                }
                return;
            }
            dfs(idx + 1, orVal | nums[idx]);
            dfs(idx + 1, orVal);
        };
        dfs(0,0);
        return cnt;
    }
};

File: 3000/LC 2044.go
func countMaxOrSubsets(nums []int) (ans int) {
    maxOr := 0
    n := len(nums)
    f := make([]int, 1 << n)
    for i, x := range nums {
        maxOr |= x
        f[1<<i] = x
    }
    

    for i := 1; i < (1 << n); i++ {
        x := i & (-i)
        if x ^ i != 0 {
            f[i] = f[i ^ x] | f[x]
        }
        if f[i] == maxOr {
            ans++
        }
    }



    return
}